modules/module/LoRAModule.py

```diff
@@ -1,201 +1,388 @@
+        # They definitely exist at this point in the execution.
+        assert self.orig_forward
+
+        A = self.lora_down.weight
+        B = self.lora_up.weight
+        WP = self.orig_module.weight + (self.make_weight(A, B) * (self.alpha / self.rank))
+        # A norm should never really end up zero at any point, but epsilon just
+        # to be safe if we underflow or something. Also, as per section 4.3 of
+        # the paper, we treat the norm as a constant for the purposes of
+        # backpropagation in order to save VRAM (to do this, we detach it from
+        # the gradient graph).
+        norm = WP.detach() \
+                 .transpose(0, 1) \
+                 .reshape(WP.shape[1], -1) \
+                 .norm(dim=1, keepdim=True) \
+                 .reshape(WP.shape[1], *[1] * self.dora_num_dims) \
+                 .transpose(0, 1) + torch.finfo(WP.dtype).eps
+        WP = self.dora_scale * (WP / norm)
+        # In the DoRA codebase (and thus the paper results), they perform
+        # dropout on the *input*, rather than between layers, so we duplicate
+        # that here.
+        return self.op(self.dropout(x),
+                       WP,
+                       self.orig_module.bias,
+                       **self.layer_kwargs)
+
+
+DummyLoRAModule = LoRAModule.make_dummy()
+DummyDoRAModule = DoRAModule.make_dummy()
+DummyLoHaModule = LoHaModule.make_dummy()
+
+
 class LoRAModuleWrapper:
     orig_module: nn.Module
     rank: int
+    alpha: float
 
     lora_modules: dict[str, LoRAModule]
 
     def __init__(
             self,
             orig_module: nn.Module | None,
-            rank: int,
             prefix: str,
-            alpha: float = 1.0,
+            config: TrainConfig,
             module_filter: list[str] = None,
     ):
-        super(LoRAModuleWrapper, self).__init__()
         self.orig_module = orig_module
-        self.rank = rank
         self.prefix = prefix
+        self.peft_type = config.peft_type
+        self.rank = config.lora_rank
+        self.alpha = config.lora_alpha
         self.module_filter = module_filter if module_filter is not None else []
-
-        self.lora_modules = self.__create_modules(orig_module, alpha)
-
-    def __create_modules(self, orig_module: nn.Module | None, alpha: float) -> dict[str, LoRAModule]:
+        weight_decompose = config.lora_decompose
+        if self.peft_type == PeftType.LORA:
+            if weight_decompose:
+                self.klass = DoRAModule
+                self.dummy_klass = DummyDoRAModule
+                self.additional_args = [self.rank, self.alpha]
+            else:
+                self.klass = LoRAModule
+                self.dummy_klass = DummyLoRAModule
+                self.additional_args = [self.rank, self.alpha]
+        elif self.peft_type == PeftType.LOHA:
+            self.klass = LoHaModule
+            self.dummy_klass = DummyLoHaModule
+            self.additional_args = [self.rank, self.alpha]
```

Comment:
I'd change this to a `match` over `(self.peft_type, weight_decompose)` (this way, `LoHA, True` can be disallowed instead of a back-compat break later), and use a helper class like so:

```
@dataclass
class PeftSpec:
    klass: type
    dummy_klass: type
    self.additional_args: list

Reply:
This sort of match change would be a bad user experience. The user has no way to change the value of weight_decompose when LoHa is turned on. If they had it set previously, then swap to LoHa, it will not work and to fix it they'd need to go back to LoRA, change the value, then return to LoHa. Even if we tell them that in an error message, that's not a good user experience.